doc/dev/crimson/poseidonstore.rst

```diff
@@ -0,0 +1,254 @@
+
+
+Design
+======
+.. ditaa::
+
+  +-WAL partition-|----------------------Data partition-------------------------+
+  | Sharded partition 1
+  +-----------------------------------------------------------------------------+
+  | WAL -> |      | Super block | Object meta | Allocation bitmap | Data blocks |
+  +-----------------------------------------------------------------------------+
+  | Sharded partition 2
+  +-----------------------------------------------------------------------------+
+  | WAL -> |      | Super block | Object meta | Allocation bitmap | Data blocks |
+  +-----------------------------------------------------------------------------+
+  | Sharded partition N 
+  +-----------------------------------------------------------------------------+
+  | WAL -> |      | Super block | Object meta | Allocation bitmap | Data blocks |
+  +-----------------------------------------------------------------------------+
+
+
+- WAL
+  - Log and frequently updated metadata are stored as a WAL entry in the WAL partition
+  (Can be placed on NVM)
+  - Space within the WAL partition is continually reused in a circular manner
+  - Flush the metadata if necessary
+- Write procedure for Metadata
+  - Appended at the WAL first
+  - Overwrite the metadta in the data partition when flushing
+- Write procedure for Data
+  - Overwrite the data in the data partition
+- Disk layout
+  - Object meta can embed data. For example, object_info_t can be recorded as an entry of
+    the Object meta
+  - Allocation bitmap manages the Data blocks
+  - Super block manages data partitions
+
+
+I/O procedure
+-------------
+- Write
+  - 1. Append a log entry that contains pg_log, snapset, object_infot_t, block allocation
+    using NVMe atomic write command on the WAL
+    - NVMe provides atomicity guarantees for a write command (Atomic Write Unit Power Fail)
+      For example, 512 Kbytes of data can be atomically written at once without fsync()
+    - Small size (object_info_t, snapset, collection, etc.) can be embed
+    - stage 1
+      WAL (written) --> | TxBegin A | Log Entry |
+  - 2. Then, overwrite the data to data blocks
+    - stage 2 (Updating object meta and allocation bitmap can be skipped via pre-allocation)
+      WAL --> | TxBegin A | Log Entry |  
+      Data partition (written)--> | Object Meta | Allocation bitmap | Data blocks |
+  - 3. Append TxEnd with a next log entry when doing next write
+    - stage 3
+      WAL (written) --> | TxBegin A | Log Entry | TxEnd A | TxBegin B | LogEntry |
+
+- Read
+  - Use the cached object metadata to find out the data location
+  - If not cached, need to search WAL after checkpoint and Object meta partition to find the 
+    latest meta data
+
+- Flush
+  - Flush WAL entries whenever needed
+
+
+Recovery procedure
+------------------
+- In-place update may lead to partial write to data 
+  - We can recover data from replicas in the case of inconsistency (we need to detect it)
```

Comment:
I think this is a *very* bad idea for a bunch of reasons:
1) In general, a particular object will be overwritten on all replicas at a similar point in time, so you'd have a decent chance of *no* replica having a valid copy of the object in the case of a correlated power loss (correlated power losses are common).
2) "we need to detect it" This is worse than it seems.
  - The WAL does give you an idea of where there were in-progress writes, but you then need to map those blocks onto rados level objects and then do a scan on those objects against other osds with the same PG.
  - This process would necessarily have to happen before we actually trust *any* of the data in any pg on an osd after restart.
  - It essentially amounts to a new incredibly complicated phase of peering (keep in mind, it's likely after a correlated crash that *all* osds in a PG have some objects that are invalid) where we have to do a sort of limited deep scrub before we can serve OSD.
  - Even without the complexity and correctness problems above, restarting a single OSD uncleanly would require a bunch of other osds to immediately do a bunch of reads (total IO proportional to max write rate of the osd which restarted) before the OSD in question can be allowed to serve IO.

I do not think you have fully thought through the implications of what using other replicas to heal inconsistent writes would entail.